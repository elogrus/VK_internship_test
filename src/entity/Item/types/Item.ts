// Не стал полностью описывать все существующие поля.
export interface IOwner {
    login: string;
    id: number;
    node_id: string;
    avatar_url: string;
    html_url: string;
}

export interface IItem {
    id: number;
    name: string;
    full_name: string;
    private: boolean;
    description: string;
    html_url: string;
    language: string;
    topics: string[];
    owner: IOwner;
}

export interface IItemList {
    total_count: number;
    items: IItem[];
}
