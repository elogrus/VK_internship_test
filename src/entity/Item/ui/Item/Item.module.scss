.Item {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    width: 100%;

    & .Content {
        display: flex;
        flex-direction: column;
        gap: 1em;
    }
}
