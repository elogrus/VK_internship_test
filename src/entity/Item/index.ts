import { IItem, IItemList, IOwner } from "./types/Item";
import { ItemList } from "./ui/ItemList/ItemList";
export { IItem, IItemList, IOwner, ItemList };

