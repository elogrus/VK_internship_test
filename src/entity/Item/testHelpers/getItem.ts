import { IItem } from "entity/Item/types/Item";

export const getTestItem = (id: number, name = "name"): IItem => ({
    id,
    description: "some desc",
    full_name: "fullname",
    html_url: "htmlUrl",
    language: "HTML",
    name,
    owner: {
        avatar_url: "avatarUrl",
        html_url: "ownerHtmlUrl",
        id: 12345,
        login: "login",
        node_id: "nodeId",
    },
    private: false,
    topics: ["topic1", "topic2", "topic3"],
});
