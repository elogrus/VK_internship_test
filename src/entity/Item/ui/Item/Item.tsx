import CheckIcon from "@mui/icons-material/Check";
import ClearIcon from "@mui/icons-material/Clear";
import DeleteIcon from "@mui/icons-material/Delete";
import EditIcon from "@mui/icons-material/Edit";
import { Card, CardActions, CardContent, IconButton } from "@mui/material";
import { IItem } from "entity/Item/types/Item";
import { memo, useRef, useState } from "react";
import * as cls from "./Item.module.scss";
import { LanguagePart } from "./parts/LanguagePart";
import { OwnerPart } from "./parts/OwnerPart";
import { ProjectPart } from "./parts/ProjectPart";

export interface ItemProps {
    item: IItem;
    onEditItem: (id: number, newItem: IItem) => void;
    onDeleteItem: (id: number) => void;
}
export interface IInputsRef {
    avatar?: string;
    login?: string;
    profileLink?: string;
    name?: string;
    description?: string;
    language?: string;
}

export interface PartProps extends ItemProps {
    InputsRef: ReturnType<typeof useRef<IInputsRef>>;
    isEdit: boolean;
}

export const Item = memo((props: ItemProps) => {
    const { item, onEditItem, onDeleteItem } = props;
    const [isEdit, setIsEdit] = useState(false);
    const InputsRef = useRef<IInputsRef>({});
    const onSave = () => {
        const newItem = {
            ...item,
            name: InputsRef.current?.name || item.name,
            description: InputsRef.current?.description || item.description,
            language: InputsRef.current?.language || item.language,
            owner: {
                ...item.owner,
                login: InputsRef.current?.login || item.owner.login,
                avatar_url: InputsRef.current?.avatar || item.owner.avatar_url,
                html_url: InputsRef.current?.profileLink || item.owner.html_url,
            },
        };
        onEditItem(item.id, newItem);
        setIsEdit(false);
    };

    const onStartEdit = () => {
        setIsEdit(true);
    };

    const onDelete = () => {
        onDeleteItem(item.id);
    };

    const onClear = () => {
        setIsEdit(false);
    };

    return (
        <Card data-testid="Item_root" className={cls.Item}>
            <CardContent className={cls.Content}>
                <OwnerPart
                    item={item}
                    isEdit={isEdit}
                    InputsRef={InputsRef}
                    {...props}
                />
                <ProjectPart
                    item={item}
                    isEdit={isEdit}
                    InputsRef={InputsRef}
                    {...props}
                />
                <LanguagePart
                    item={item}
                    isEdit={isEdit}
                    InputsRef={InputsRef}
                    {...props}
                />
            </CardContent>

            <CardActions data-testid="Item_buttons">
                <IconButton
                    onClick={!isEdit ? onStartEdit : onSave}
                    color="info"
                >
                    {!isEdit ? (
                        <EditIcon data-testid="Item_editButton" />
                    ) : (
                        <CheckIcon data-testid="Item_saveButton" />
                    )}
                </IconButton>
                <IconButton
                    color="error"
                    onClick={!isEdit ? onDelete : onClear}
                >
                    {!isEdit ? (
                        <DeleteIcon data-testid="Item_deleteButton" />
                    ) : (
                        <ClearIcon data-testid="Item_clearButton" />
                    )}
                </IconButton>
            </CardActions>
        </Card>
    );
});
