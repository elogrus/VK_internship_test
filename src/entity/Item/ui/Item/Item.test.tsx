import { fireEvent, render, screen } from "@testing-library/react";
import { IItem } from "entity/Item/types/Item";
import GithubLanguageColors from "shared/consts/GithubLanguageColors";
import { getTestItem } from "../../testHelpers/getItem";
import { Item } from "./Item";

const itemProp = getTestItem(123, "name");

const itemPropWithoutLanguage: IItem = {
    ...itemProp,
    language: "",
};

const onEditItem = jest.fn();
const onDeleteItem = jest.fn();

describe("Item", () => {
    const renderItem = (item: IItem = itemProp) => {
        render(
            <Item
                item={item}
                onEditItem={onEditItem}
                onDeleteItem={onDeleteItem}
            />
        );
    };

    const checkExistInputs = (exists: boolean = true) => {
        if (exists) {
            expect(
                screen.queryByPlaceholderText("Ссылка на аватар")
            ).toBeInTheDocument();
            expect(screen.queryByPlaceholderText("Логин")).toBeInTheDocument();
            expect(
                screen.queryByPlaceholderText("Ссылка на профиль")
            ).toBeInTheDocument();
            expect(
                screen.queryByPlaceholderText("Название")
            ).toBeInTheDocument();
            expect(
                screen.queryByPlaceholderText("Описание")
            ).toBeInTheDocument();
            expect(
                screen.queryByPlaceholderText("Язык программирования")
            ).toBeInTheDocument();
        } else {
            expect(
                screen.queryByPlaceholderText("Ссылка на аватар")
            ).toBeNull();
            expect(screen.queryByPlaceholderText("Логин")).toBeNull();
            expect(
                screen.queryByPlaceholderText("Ссылка на профиль")
            ).toBeNull();
            expect(screen.queryByPlaceholderText("Название")).toBeNull();
            expect(screen.queryByPlaceholderText("Описание")).toBeNull();
            expect(
                screen.queryByPlaceholderText("Язык программирования")
            ).toBeNull();
        }
    };

    const changeInputs = () => {
        const loginInput = screen.queryByPlaceholderText("Логин");
        const avatarUrlInput =
            screen.queryByPlaceholderText("Ссылка на аватар");
        const profileUrlInput =
            screen.queryByPlaceholderText("Ссылка на профиль");
        const nameInput = screen.queryByPlaceholderText("Название");
        const descInput = screen.queryByPlaceholderText("Описание");
        const languageInput = screen.queryByPlaceholderText(
            "Язык программирования"
        );

        fireEvent.change(loginInput, { target: { value: "новый логин" } });
        fireEvent.change(avatarUrlInput, {
            target: { value: "новая ссылка на аватар" },
        });
        fireEvent.change(profileUrlInput, {
            target: { value: "новая ссылка на профиль" },
        });
        fireEvent.change(nameInput, { target: { value: "новое название" } });
        fireEvent.change(descInput, { target: { value: "новое описание" } });
        fireEvent.change(languageInput, {
            target: { value: "новый язык программирования" },
        });

        const newItem: IItem = {
            ...itemProp,
            owner: { ...itemProp.owner },
        };
        newItem.owner.login = "новый логин";
        newItem.owner.avatar_url = "новая ссылка на аватар";
        newItem.owner.html_url = "новая ссылка на профиль";
        newItem.name = "новое название";
        newItem.description = "новое описание";
        newItem.language = "новый язык программирования";

        return newItem;
    };

    it("renders", () => {
        renderItem();
        const root = screen.queryByTestId("Item_root");
        expect(root).toBeInTheDocument();
        expect(root).toMatchSnapshot();
    });

    it("shows owner info", () => {
        renderItem();

        const owner = screen.queryByTestId("Item_owner");
        expect(owner).toBeInTheDocument();
        expect(owner).toHaveAttribute("href", "ownerHtmlUrl");

        expect(screen.queryByText("login")).toBeInTheDocument();

        const avatar = screen.queryByTestId("Item_avatar");
        expect(avatar).toBeInTheDocument();
        expect(avatar.querySelector("img")).toHaveAttribute("src", "avatarUrl");
    });

    it("shows project info", () => {
        renderItem();

        const project = screen.queryByTestId("Item_project");
        expect(project).toBeInTheDocument();
        expect(project).toHaveAttribute("href", "htmlUrl");

        expect(screen.queryByText("name")).toBeInTheDocument();
        expect(screen.queryByText("some desc")).toBeInTheDocument();
    });

    it("shows language info", () => {
        renderItem();
        expect(screen.queryByText("HTML")).toBeInTheDocument();
        const colorCircle = screen.queryByTestId("Item_colorCircle");
        expect(colorCircle).toBeInTheDocument();
        expect(colorCircle).toHaveStyle(
            `backgroundColor: ${GithubLanguageColors["HTML"]}`
        );
    });

    it("don't shows language info with empty language", () => {
        renderItem(itemPropWithoutLanguage);
        expect(screen.queryByText("HTML")).toBeNull();
    });

    it("show buttons in not-edit mode", () => {
        render(
            <Item
                item={itemProp}
                onEditItem={onEditItem}
                onDeleteItem={onDeleteItem}
            />
        );
        const buttons = screen.queryByTestId("Item_buttons");

        expect(
            buttons.querySelector("*[data-testid=Item_editButton]")
        ).toBeInTheDocument();
        expect(
            buttons.querySelector("*[data-testid=Item_deleteButton]")
        ).toBeInTheDocument();

        expect(
            buttons.querySelector("*[data-testid=Item_saveButton]")
        ).toBeNull();
        expect(
            buttons.querySelector("*[data-testid=Item_clearButton]")
        ).toBeNull();
    });

    it("show buttons in edit mode", () => {
        render(
            <Item
                item={itemProp}
                onEditItem={onEditItem}
                onDeleteItem={onDeleteItem}
            />
        );
        const buttons = screen.queryByTestId("Item_buttons");
        const editButton = screen.queryByTestId("Item_editButton");
        fireEvent.click(editButton);
        expect(
            buttons.querySelector("*[data-testid=Item_editButton]")
        ).toBeNull();
        expect(
            buttons.querySelector("*[data-testid=Item_deleteButton]")
        ).toBeNull();

        expect(
            buttons.querySelector("*[data-testid=Item_saveButton]")
        ).toBeInTheDocument();
        expect(
            buttons.querySelector("*[data-testid=Item_clearButton]")
        ).toBeInTheDocument();
    });

    it("show inputs in edit mode", () => {
        render(
            <Item
                item={itemProp}
                onEditItem={onEditItem}
                onDeleteItem={onDeleteItem}
            />
        );
        const editButton = screen.queryByTestId("Item_editButton");
        checkExistInputs(false);
        fireEvent.click(editButton);
        checkExistInputs(true);
    });

    it("onDelete", () => {
        render(
            <Item
                item={itemProp}
                onEditItem={onEditItem}
                onDeleteItem={onDeleteItem}
            />
        );
        fireEvent.click(screen.queryByTestId("Item_deleteButton"));
        expect(onDeleteItem).toHaveBeenCalledTimes(1);
        expect(onDeleteItem).toHaveBeenCalledWith(123);
    });

    it("on change data", () => {
        render(
            <Item
                item={itemProp}
                onEditItem={onEditItem}
                onDeleteItem={onDeleteItem}
            />
        );
        fireEvent.click(screen.queryByTestId("Item_editButton"));
        const newItem = changeInputs();
        fireEvent.click(screen.queryByTestId("Item_saveButton"));
        expect(onEditItem).toHaveBeenCalledTimes(1);
        expect(onEditItem).toHaveBeenCalledWith(123, newItem);
    });

    it("on clear don't change data", () => {
        render(
            <Item
                item={itemProp}
                onEditItem={onEditItem}
                onDeleteItem={onDeleteItem}
            />
        );
        fireEvent.click(screen.queryByTestId("Item_editButton"));
        changeInputs();
        fireEvent.click(screen.queryByTestId("Item_clearButton"));
        expect(onEditItem).toHaveBeenCalledTimes(0);
    });
});
