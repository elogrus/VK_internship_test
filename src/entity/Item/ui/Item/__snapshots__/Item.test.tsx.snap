// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
  data-testid="Item_root"
  style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
>
  <div
    class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
  >
    <div
      class="MuiBox-root css-0"
    >
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
        data-testid="Item_owner"
        href="ownerHtmlUrl"
        rel="noreferrer"
        target="_blank"
      >
        <div
          aria-label="login"
          class="MuiAvatar-root MuiAvatar-circular css-fmeqll-MuiAvatar-root"
          data-testid="Item_avatar"
        >
          <img
            class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
            src="avatarUrl"
          />
        </div>
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
        >
          login
        </h6>
      </a>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-18rqtra-MuiTypography-root-MuiLink-root"
        data-testid="Item_project"
        href="htmlUrl"
        rel="noreferrer"
        target="_blank"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
        >
          name
        </h5>
        <span
          class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
        >
          some desc
        </span>
      </a>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2 css-1ypmsg6-MuiTypography-root"
      >
        <div
          data-testid="Item_colorCircle"
          style="background-color: rgb(227, 76, 38);"
        />
        HTML
      </h6>
    </div>
  </div>
  <div
    class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
    data-testid="Item_buttons"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInfo MuiIconButton-sizeMedium css-1mjraff-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="Item_editButton"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
        />
      </svg>
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium css-lky15p-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="Item_deleteButton"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
        />
      </svg>
    </button>
  </div>
</div>
`;
